module/VuFind/src/VuFind/Db/Service/LoginTokenService.php

```diff
@@ -148,7 +185,24 @@ public function deleteByUser(UserEntityInterface|int $userOrId): void
     public function getByUser(UserEntityInterface|int $userOrId, bool $grouped = true): array
     {
         $userId = is_int($userOrId) ? $userOrId : $userOrId->getId();
-        return $this->getDbTable('LoginToken')->getByUserId($userId, $grouped);
+        $dql = 'SELECT lt '
+            . 'FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' lt '
+            . 'WHERE lt.userId = :userId '
+            . 'ORDER BY lt.lastLogin DESC';
+
+        if ($grouped) {
+            // Modify the DQL for grouping logic
+            $dql = 'SELECT MAX(lt.lastLogin) as lastLogin, lt.series, lt.browser, lt.platform, lt.expires '
```

Comment:
Does this query result in an array of LoginTokenEntityInterface objects being returned? I'm wondering if it might instead be resulting in a simple array, which might then fail to match the expected return value of this method. I can't remember exactly what this option is for, but let me know if you need help digging into this further.

Reply:
I think this comment is still a problem: when I log in and look at my user profile page, I get this fatal error:

`Call to a member function getPlatform() on array`

I think this is because the grouped query is returning an array of arrays, but the calling code is expecting an array of LoginTokenEntityInterface objects.

I think we need a way to retrieve the actual entities that represent the groups in the grouped query. I'm not exactly sure how to do this, though! Maybe there's some clever way to extract IDs from the group, and then use a subquery to fetch those specific rows by ID from the database... but I'm not sure off the top of my head what SQL (or, more importantly, DQL) mechanism would let us pick the ID from the group that corresponds with the most recent login date. Or maybe we need to do this as a multi-step process, which is not ideal, but may be okay given that this is likely to be called infrequently and on relatively small sets of data.

If you don't have any ideas about how to proceed, maybe we can brainstorm with others in the community -- let me know and I can reach out.

Other options to consider if we can't find a better solution:

1.) Create a GroupedLoginTokenEntity class that can implement LoginTokenEntityInterface so that we can wrap the arrays returned here in an object format that the templates can understand. Feels clunky but it would at least maintain the right interfaces.

2.) Make the template tolerant of arrays as well as entity objects. This is the easiest solution, but I would much rather maintain consistent interfaces if we can!

What do you think?